### DISCUSSION TITLE
How to implement a temperature-dependent rank 2 tensor.

### DISCUSSION THREAD

**Discussion by marinsiebert** [idaholab/moose#26116] [Q&A General] [answered]
Hi,
so basically what the titel says, I want to have a rank 2 tensor with a temperature dependent value along the diagonal for the thermal conductivity. What would be the easiest way to achieve this? 
The goal would be to have an additional PorousFlowThermalConductivity Material that can calculate the thermal conductivity of a porous material depending on the porosity, saturation and the temperature-depending thermal conductivities of the porefluid (water) and the granular matrix. Find below the current state my .C file is in

`//* This file is part of the MOOSE framework
//* https://www.mooseframework.org
//*
//* All rights reserved, see COPYRIGHT for full restrictions
//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html

#include "PorousFlowThermalConductivityBrakelmann.h"

registerMooseObject("PorousFlowApp", PorousFlowThermalConductivityBrakelmann);

InputParameters
PorousFlowThermalConductivityFromPorosity::validParams()
{
  InputParameters params = PorousFlowThermalConductivityBase::validParams();
  paras.addRequiredCoupledVar("temperature");
  params.addRequiredParam<Real>"rho_s",2650,
                                          "The density of the soil mineral solids");

  params.addRequiredParam<RealTensorValue>("lambda_b",
                                           "The thermal conductivity of soil mineral solids approximated by"
                                           "0.0812*sand_content+0.054*silt_content+0.02*clay_content");
  params.addRequiredParam<RealTensorValue>("lambda_w,"
                                           "The approximated temperature-depending thermal conductivity of water"
                                           "(e.g. Kays et al.,2005): -8.354e-6T^2+6.53e-3T-0.5981");
  params.addParam<unsigned>("aqueous_phase_number",
                             0,
                             "The phase number of the aqueous phase.");
  params.addClassDescription("This Material calculates rock-fluid combined thermal conductivity "
                             "for the unsaturated porous medium using a empirical"
                             "non-linear whole-range function from Brakelmann (2015)."
                             "Thermal conductivity = lambda_w^(phi)*lambda_b^(1-phi)*exp(-3.08*phi*(1-S)^2)) "
                             "where phi is porosity, lambda_w, lambda_b are "
                             "thermal conductivities of the fluid and solid"
                             "(approximated from grain size distribution)"
                              "and S is the degree of saturation"
                            ););
  return params;
}

PorousFlowThermalConductivityBrakelmann::PorousFlowThermalConductivityBrakelmann(
    const InputParameters & parameters)
  : PorousFlowThermalConductivityBase(parameters),
    _la_b(getParam<RealTensorValue>)("lambda_b")),
    _la_w(getParam<RealTensorValue>)("lamda_w")),
    _rho_s(getParam<Real>("rho_s")),
    _rho_b(getParam<Real>("rho_b")),
    _porosity_qp(getMaterialProperty<Real>("PorousFlow_porosity_qp")),
    _dporosity_qp_dvar(getMaterialProperty<std::vector<Real>>("dPorousFlow_porosity_qp_dvar"))
    _aqueous_phase(_num_phases > 0),
    _aqueous_phase_number(getParam<unsigned>("aqueous_phase_number")),
    _saturation_qp(_aqueous_phase
                       ? &getMaterialProperty<std::vector<Real>>("PorousFlow_saturation_qp")
                       : nullptr),
    _dsaturation_qp_dvar(_aqueous_phase ? &getMaterialProperty<std::vector<std::vector<Real>>>(
                                              "dPorousFlow_saturation_qp_dvar")
                                        : nullptr)
{
  if (_num_phases != 1)
    paramError("fluid_phase",
               "The Dictator proclaims that the number of phases is ",
               _dictator.numPhases(),
               " whereas this material can only be used for single phase "
               "simulations.  Be aware that the Dictator has noted your mistake.");
}

void
PorousFlowThermalConductivityBrakelmann::computeQpProperties()
{
  _la_qp[_qp] =
    std::pow(_la_w, _porosity_qp)*pow(_la_b, 1-_porosity_qp)*
    exp(-3.08*_porosity_qp*pow((1-((*_saturation_qp)[_qp][_aqeous_phase_number])/_porosity_qp),2))
}
`

**Comment by GiudGiud** [comment]
Hello

Are the two diagonal coefficients going to have the same temperature dependence?

Do you want to hard code the dependence of the coefficients or do you want to be able to change it from user input file? 

Guillaume

**Reply by marinsiebert** [reply]
For now I am fine if the diagonal has the same results.

The thermal conductivity for water will be calculated by  -8.354e-6T^2+6.53e-3T-0.5981, this doesnt necessarily have to be subject to user input, however now thinking about it it probably wont hurt in the long term to either allow users to be able to change the approximation in the input file or, even better, let the material consume the thermal conductivities as calculated by the Water97FluidProperties object.

The thermal conductivity for the solid material will be 0.0812*sand_content+0.054*silt_content+0.02*clay_content - this does not necessarily have to be calculated in the material, as the users could just calculate it themselves and enter it into a Parameter like has been done in the PorousFlowThermalConductivityIdeal.

**Reply by GiudGiud** [reply, marked as answer]
If you are planning to use a FluidProperties user object you should look at how the 
https://mooseframework.inl.gov/source/materials/GeneralFluidProps.html
material is implemented. 

It will be very similar for you except you ll need to define a tensor material property. 
An example of this is this one:
https://mooseframework.inl.gov/source/materials/ADGenericFunctionRankTwoTensor.html

**Reply by marinsiebert** [reply]
Tank you, that is exactly what I was looking for!
